src/common/hive_partitioning.cpp

```diff
@@ -126,8 +131,120 @@ void HivePartitioning::ApplyFiltersToFileList(ClientContext &context, vector<str
+	Vector hashes(LogicalType::HASH, input.size());
+	input.Hash(group_by_columns, hashes);
+
+	for (idx_t i = 0; i < input.size(); i++) {
+		HivePartitionKey key;
+		key.hash = FlatVector::GetData<hash_t>(hashes)[i];
+		for (auto &col : group_by_columns) {
+			key.values.emplace_back(input.GetValue(col, i));
+		}
+
+		auto lookup = local_partition_map.find(key);
+		const auto partition_indices = FlatVector::GetData<idx_t>(state.partition_indices);
+		if (lookup == local_partition_map.end()) {
+			idx_t new_partition_id = RegisterNewPartition(key, state);
+			partition_indices[i] = new_partition_id;
+		} else {
+			partition_indices[i] = lookup->second;
+		}
+	}
+}
+
+std::map<idx_t, const HivePartitionKey *> HivePartitionedColumnData::GetReverseMap() {
+	std::map<idx_t, const HivePartitionKey *> ret;
+	for (const auto &pair : local_partition_map) {
+		ret[pair.second] = &(pair.first);
+	}
+	return ret;
+}
+
+void HivePartitionedColumnData::GrowAllocators() {
+	idx_t current_allocator_size = allocators->allocators.size();
+	idx_t required_allocators = local_partition_map.size();
+
+	allocators->allocators.reserve(current_allocator_size);
+	for (idx_t i = current_allocator_size; i < required_allocators; i++) {
+		CreateAllocator();
+	}
+
+	D_ASSERT(allocators->allocators.size() == local_partition_map.size());
+}
+
+void HivePartitionedColumnData::GrowAppendState(PartitionedColumnDataAppendState &state) {
+	idx_t current_append_state_size = state.partition_append_states.size();
+	idx_t required_append_state_size = local_partition_map.size();
+
+	for (idx_t i = current_append_state_size; i < required_append_state_size; i++) {
+		state.partition_append_states.emplace_back(make_unique<ColumnDataAppendState>());
+		state.partition_buffers.emplace_back(CreatePartitionBuffer());
+	}
+}
+
+void HivePartitionedColumnData::GrowPartitions(PartitionedColumnDataAppendState &state) {
+	idx_t current_partitions = partitions.size();
+	idx_t required_partitions = local_partition_map.size();
+
+	D_ASSERT(allocators->allocators.size() == required_partitions);
+
+	for (idx_t i = current_partitions; i < required_partitions; i++) {
+		partitions.emplace_back(CreatePartitionCollection(i));
+		partitions[i]->InitializeAppend(*state.partition_append_states[i]);
+	}
+	D_ASSERT(partitions.size() == local_partition_map.size());
+}
+
+void HivePartitionedColumnData::SynchronizeLocalMap() {
+	// Synchronise global map into local, may contain changes from other threads too
+	for (auto it = global_state->partitions.begin() + local_partition_map.size(); it < global_state->partitions.end();
+	     it++) {
+		local_partition_map[(*it)->first] = (*it)->second;
+	}
+}
+
+idx_t HivePartitionedColumnData::RegisterNewPartition(HivePartitionKey key, PartitionedColumnDataAppendState &state) {
+	if (global_state) {
+		// We need to lock both the GlobalHivePartitionState and the allocators while adding a partition
+		unique_lock<mutex> lck_gstate(global_state->lock);
```

Comment:
Do we need to hold two separate locks at the same time here? Using multiple locks is quite error-prone due to lock inversion.

Reply:
Good one, i took a better look at the locking here, and we definitely don't need to hold two locks, changed this to:

- acquire global state lock to sync global to local and vice versa
- release lock
- grow allocators (with allocator lock) to the required size for this thread. This is enough to guarantee there are always enough allocators available for each thread